From 201ceadd26dee2ed8f01613fd3a42bddfdccd80b Mon Sep 17 00:00:00 2001
From: Shawn Landden <shawn@churchofgit.com>
Date: Mon, 16 Dec 2013 13:52:13 -0800
Subject: [PATCH] ssl: multiplex ssh only ssl connections

Would be nice to properly integrate this with the nginx sources
including a compile time option and a configuration directive to
listen like spdy has.
---
 src/http/ngx_http_request.c | 36 +++++++++++++++++++++++++++++++++++-
 1 file changed, 35 insertions(+), 1 deletion(-)

diff --git a/src/http/ngx_http_request.c b/src/http/ngx_http_request.c
index c8c5d15..24c9e29 100644
--- a/src/http/ngx_http_request.c
+++ b/src/http/ngx_http_request.c
@@ -9,6 +9,10 @@
 #include <ngx_core.h>
 #include <ngx_http.h>
 
+#if 1 //NGX_HAVE_SSH_PASS
+# include <unistd.h>
+# include <fcntl.h>
+#endif
 
 static void ngx_http_wait_request_handler(ngx_event_t *ev);
 static void ngx_http_process_request_line(ngx_event_t *rev);
@@ -696,7 +700,37 @@ ngx_http_ssl_handshake(ngx_event_t *rev)
 
             return;
         }
-
+#if 1 //NGX_HAVE_SSH_PASS
+        do {
+            u_char sb[4];
+
+            n = recv(c->fd, (char *) sb, 4, MSG_PEEK);
+            if (n == 4)
+                if (sb[0] == 'S' && sb[1] == 'S' && sb[2] == 'H' && sb[3] == '-') {
+                    int r;
+                    pid_t pid;
+
+                    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, rev->log, 0, "ssh");
+                    if ((pid = fork()) == 0) {
+                        if (c->fd != 3) {
+                            r = dup2(c->fd, 3);
+                            if (r < 0)
+                                _exit(1);
+                            close(c->fd);
+                        }
+                        r = fcntl(3, F_SETFD, 0); /* remove any FD_CLOEXEC flag */
+                        if (r < 0)
+                            _exit(1);
+                        r = execve("/usr/lib/nginx/spawn-ssh-suid",
+                                   (char *[]){"spawn-ssh-suid", NULL},
+                                   (char *[]){NULL});
+                        _exit(1);
+                    }
+                    ngx_close_connection(c);
+                    return;
+                }
+        } while (0);
+#endif
         ngx_log_debug0(NGX_LOG_DEBUG_HTTP, rev->log, 0, "plain http");
 
         c->log->action = "waiting for request";
-- 
1.8.5.1

